daemon: validate hostname length to provide better error message

majiayu000 opened the pull request:
## Summary

Add validation for hostname length during container creation to provide a more user-friendly error message.

## Problem

Linux limits hostnames to 64 bytes (HOST_NAME_MAX). When users create containers with hostnames exceeding this limit, they receive a cryptic error:

```
OCI runtime create failed: ... sethostname: invalid argument
```

This is particularly problematic when using Swarm hostname templates that expand to long strings.

## Solution

Add hostname length validation in `validateContainerConfig()`. Now users will see:

```
hostname "..." is too long (max 64 characters)
```

## Changes

- Added `maxHostnameLen` constant (64 bytes, matching Linux HOST_NAME_MAX)
- Added hostname length validation in `validateContainerConfig()`
- Added unit test for hostname length validation

Fixes #36402

majiayu000 commented on daemon/container.go:
```diff
@@ -232,6 +232,10 @@ func (daemon *Daemon) verifyContainerSettings(daemonCfg *configStore, hostConfig
 
+// maxHostnameLen is the maximum length of a hostname on Linux.
+// This is defined by HOST_NAME_MAX in the kernel.
+const maxHostnameLen = 64
```
Thanks for the feedback. This change is intentionally aligned with Linux HOST_NAME_MAX (64 bytes), so I kept the limit at 64 and clarified the error to say "bytes". The DNS label 63-byte limit seems specific to alias/DNS registration, so I'd prefer to enforce that in the network/alias validation path rather than tightening the hostname limit here. Let me know if you'd like me to add a separate alias/DNS check.

thaJeztah replied:
> 64 includes the null terminator, so the max should be 63 anyway.

So, that's what I thought as well, but it looks like it's really 64, so not "including the nul-terminator";

```bash
docker run --rm --hostname '0123456789012345678901234567890123456789012345678901234567891234' busybox hostname
0123456789012345678901234567890123456789012345678901234567891234

docker run --rm --hostname '01234567890123456789012345678901234567890123456789012345678912345' busybox hostname
docker: Error response from daemon: failed to create task for container: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: sethostname: invalid argument

Run 'docker run --help' for more information
```

> So the constant should probably be defined in container_linux.go and container_windows.go with different values - or, perhaps better unless we can be sure of the rules for Windows, the validation should be moved to the non-Windows implementation of verifyPlatformContainerSettings (in daemon_unix.go).

Yes, I was considering that as well. And I think this has been (at least one of) the reason for not applying validation on our side, because what's supported depends on the platform and runtime used, so unless there's a strict limit defined (which could be as part of the OCI runtime spec), we try to keep our side of things agnostic.

I recall we once had something similar with validation of env-var names, where we were too strict, and as a result invalidating env-var names that would be accepted by the platform.

thaJeztah commented:
> So, I don't understand why (!),

Digging a bit, I think it may be `__NEW_UTS_LEN`, and that looks to be defined as 64

https://github.com/torvalds/linux/blob/3609fa95fb0f2c1b099e69e56634edb8fc03f87c/include/uapi/linux/utsname.h#L15

But then gets a "+1" to account for the nil-terminator;

https://github.com/torvalds/linux/blob/3609fa95fb0f2c1b099e69e56634edb8fc03f87c/include/uapi/linux/utsname.h#L31

thaJeztah commented:
Looks like at least one commit is missing a DCO sign-off (causing CI to fail). I also see that there's one commit using a different name;


```
Signed-off-by: majiayu000 <1835304752@qq.com>
Signed-off-by: apple <1835304752@qq.com>
```

Could you;

- Squash the commits
- Use your real name (instead of GitHub handle) for the sign-off
- Oh! And I see the first commit message has a `Fixes #36402` in the commit message; we try to avoid those in the commit message to prevent noise (GitHub can be noisy when including it in the commit message), and to prevent GitHub from closing things too soon (which sometimes happens)

majiayu000 committed:
daemon: validate hostname length to provide better error message

Previously, users would see a cryptic error like:
  OCI runtime create failed: ... sethostname: invalid argument

Now they will see a clear error message:
  hostname "..." is too long (max 64 characters)

This validation catches the error early during container creation,
providing a much more user-friendly error message.

Signed-off-by: Jiayu Ma <1835304752@qq.com>

majiayu000 commented:
@thaJeztah Thank you for the review! I've addressed your feedback:

Squashed the commits.
Updated the sign-off to use my real name (Jiayu Ma).
Removed the Fixes #36402 from the commit message.

Please let me know if there is anything else.
